File: api/src/main/java/com/github/noonmaru/tap/event/EntityEventManager.java
/*
 *
 *  * Copyright (c) 2020 Noonmaru
 *  *
 *  * Licensed under the General Public License, Version 3.0 (the "License");
 *  *  you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  * https://opensource.org/licenses/gpl-3.0
 *  *
 *  *  Unless required by applicable law or agreed to in writing, software
 *  *  distributed under the License is distributed on an "AS IS" BASIS,
 *  *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  See the License for the specific language governing permissions and
 *  *  limitations under the License.
 *
 */

package com.github.noonmaru.tap.event;

import com.google.common.base.Preconditions;
import com.google.common.collect.MapMaker;
import org.bukkit.entity.Entity;
import org.bukkit.event.Event;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.plugin.EventExecutor;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Map;

/**
 * @author Nemo
 */
public final class EntityEventManager {

    private final Plugin plugin;

    private final EventPriority priority;

    private final HashMap<Class<?>, ListenerStatement> statements = new HashMap<>();

    private final HashMap<Class<?>, EventListener> listeners = new HashMap<>();

    private final Map<Entity, EventEntity> entities = new MapMaker().weakKeys().makeMap();

    private final EventExecutor eventExecutor = (listener, event) -> ((EventListener) listener).onEvent(event);

    public EntityEventManager(@NotNull final Plugin plugin) {
        this(plugin, EventPriority.NORMAL);
    }

    public EntityEventManager(@NotNull final Plugin plugin, @NotNull final EventPriority priority) {
        this.plugin = plugin;
        this.priority = priority;
    }

    @NotNull
    public RegisteredEntityListener registerEvents(@NotNull final Entity entity, @NotNull final Listener listener) {

        Preconditions.checkArgument(entity.isValid(), "Invalid entity: " + entity);

        ListenerStatement listenerStatement = createRegisteredListenerStatement(listener.getClass());
        EventEntity eventEntity = entities.computeIfAbsent(entity, target -> new EventEntity());

        RegisteredEntityListener registeredEntityListener = new RegisteredEntityListener(eventEntity, listenerStatement, listener);
        eventEntity.register(registeredEntityListener);

        return registeredEntityListener;
    }

    @NotNull
    private ListenerStatement createRegisteredListenerStatement(@NotNull final Class<?> listenerClass) {
        return statements.computeIfAbsent(listenerClass, clazz -> {
            ListenerStatement statement = ListenerStatement.getOrCreate(clazz);

            for (HandlerStatement statementStatement : statement.getHandlerStatements()) {
                registerEvent(statementStatement);
            }

            return statement;
        });
    }

    private void registerEvent(@NotNull final HandlerStatement statement) {
        Class<?> registrationClass = statement.getRegistrationClass();

        EventListener listener = listeners.computeIfAbsent(registrationClass, clazz -> {
            EventListener newListener = new EventListener();
            plugin.getServer().getPluginManager().registerEvent(clazz.asSubclass(Event.class), newListener, priority, eventExecutor, plugin, false);

            return newListener;
        });

        listener.addProvider(statement.getProvider());
    }

    public void unregisterEvent(@NotNull final Entity entity, @NotNull final Listener listener) {
        EventEntity eventEntity = entities.get(entity);

        if (eventEntity != null) {
            ListenerStatement statement = statements.get(listener.getClass());

            if (statement != null) {
                eventEntity.unregister(statement, listener);
            }
        }
    }

    private class EventListener implements Listener {
        private final LinkedHashSet<EventEntityProvider> providers = new LinkedHashSet<>();

        private EventEntityProvider[] bake;

        @SuppressWarnings("unchecked")
        void onEvent(@NotNull final Event event) {
            for (EventEntityProvider provider : getBake()) {
                Class<?> eventClass = event.getClass();

                if (provider.getEventClass().isAssignableFrom(eventClass)) {
                    Entity entity = provider.getProvider().getFrom(event);

                    if (entity != null) {
                        EventEntity eventEntity = entities.get(entity);

                        if (eventEntity != null) {
                            Class<?> regClass = EventTools.getRegistrationClass(eventClass);
                            EntityHandlerList handlers = eventEntity.getHandlerList(regClass);

                            if (handlers != null)
                                handlers.callEvent(event, provider);
                        }
                    }
                }
            }
        }

        @NotNull
        private EventEntityProvider[] getBake() {
            EventEntityProvider[] bake = this.bake;

            if (bake != null)
                return bake;

            return this.bake = providers.toArray(new EventEntityProvider[0]);
        }

        private void addProvider(@NotNull final EventEntityProvider provider) {
            this.providers.add(provider);
            this.bake = null;
        }
    }

}

File: api/src/main/kotlin/com/github/noonmaru/tap/event/EntityEventManager.kt
/*
 * Copyright (c) 2020 Noonmaru
 *
 *  Licensed under the General Public License, Version 3.0 (the "License");
 *  you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/gpl-3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.noonmaru.tap.event

import com.github.noonmaru.tap.event.EventTools.getRegistrationClass
import com.google.common.collect.MapMaker
import org.bukkit.entity.Entity
import org.bukkit.event.Event
import org.bukkit.event.EventPriority
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.plugin.EventExecutor
import org.bukkit.plugin.Plugin

@Suppress("unused")
class EntityEventManager @JvmOverloads constructor(private val plugin: Plugin, private val priority: EventPriority = EventPriority.NORMAL) {
    private val statements = HashMap<Class<*>, ListenerStatement>()

    private val listeners = HashMap<Class<*>, EventListener>()

    private val entities = MapMaker().weakKeys().makeMap<Entity, EventEntity>()

    private val eventExecutor = EventExecutor { listener: Listener, event: Event ->
        (listener as EventListener).onEvent(event)
    }

    fun registerEvents(entity: Entity, listener: Listener): RegisteredEntityListener {
        require(entity.isValid) { "Invalid entity: $entity" }

        val listenerStatement = createRegisteredListenerStatement(listener.javaClass)
        val eventEntity = entities.computeIfAbsent(entity) { EventEntity() }

        val registeredEntityListener = RegisteredEntityListener(eventEntity, listenerStatement, listener)
        eventEntity.register(registeredEntityListener)

        return registeredEntityListener
    }

    private fun createRegisteredListenerStatement(listenerClass: Class<*>): ListenerStatement {
        return statements.computeIfAbsent(listenerClass) { clazz: Class<*> ->
            val statement = ListenerStatement.getOrCreate(clazz)

            for (statementStatement in statement.handlerStatements) {
                registerEvent(statementStatement)
            }

            statement
        }
    }

    private fun registerEvent(statement: HandlerStatement) {
        val registrationClass = statement.registrationClass

        val listener = listeners.computeIfAbsent(registrationClass) { clazz: Class<*> ->
            val newListener = EventListener()
            plugin.server.pluginManager.registerEvent(clazz.asSubclass(Event::class.java), newListener, priority, eventExecutor, plugin, false)

            newListener
        }

        listener.addProvider(statement.provider)
    }

    fun unregisterEvent(entity: Entity, listener: Listener) {
        val eventEntity = entities[entity]

        if (eventEntity != null) {
            val statement = statements[listener.javaClass]

            if (statement != null) {
                eventEntity.unregister(statement, listener)
            }
        }
    }

    fun unregisterAll() {
        for (eventEntity in entities.values) {
            eventEntity.unregisterAll()
        }

        for (listener in listeners.values) {
            HandlerList.unregisterAll(listener)
        }

        entities.clear()
        listeners.clear()
        statements.clear()
    }

    private inner class EventListener : Listener {
        private val providers = LinkedHashSet<EventEntityProvider>()

        private var bake: Array<EventEntityProvider>? = null

        fun onEvent(event: Event) {
            for (provider in getBake()) {
                val eventClass = event.javaClass

                if (provider.eventClass.isAssignableFrom(eventClass)) {
                    val entity = provider.provider.getFrom(event)

                    if (entity != null) {
                        val eventEntity = entities[entity]

                        if (eventEntity != null) {
                            val regClass = getRegistrationClass(eventClass)
                            val handlers = eventEntity.getHandlerList(regClass)

                            handlers?.callEvent(event, provider)
                        }
                    }
                }
            }
        }

        private fun getBake(): Array<EventEntityProvider> {
            val bake = bake
            return bake ?: providers.toTypedArray().also { this.bake = it }
        }

        fun addProvider(provider: EventEntityProvider) {
            providers.add(provider)
            bake = null
        }
    }
}